soc/xtensa/intel_adsp/common/include/cpu_init.h

```diff
@@ -0,0 +1,94 @@
+/* Copyright (c) 2021 Intel Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef __INTEL_ADSP_CPU_INIT_H
+#define __INTEL_ADSP_CPU_INIT_H
+
+#define CxL1CCAP (*(volatile uint32_t *)0x9F080080)
+#define CxL1CCFG (*(volatile uint32_t *)0x9F080084)
+#define CxL1PCFG (*(volatile uint32_t *)0x9F080088)
+
+/* "Data/Instruction Cache Memory Way Count" fields */
+#define CxL1CCAP_DCMWC ((CxL1CCAP >> 16) & 7)
+#define CxL1CCAP_ICMWC ((CxL1CCAP >> 20) & 7)
+
+/* Low-level CPU initialization.  Call this immediately after entering
+ * C code to initialize the cache, protection and synchronization
+ * features.
+ */
+static ALWAYS_INLINE void cpu_early_init(void)
+{
+	uint32_t reg;
+
+	if (IS_ENABLED(CONFIG_SOC_SERIES_INTEL_CAVS_V25)) {
+		/* First, on cAVS 2.5 we need to power the cache SRAM banks
+		 * on!  Write a bit for each cache way in the bottom half of
+		 * the L1CCFG register and poll the top half for them to turn
+		 * on.
+		 */
+		uint32_t dmask = BIT(CxL1CCAP_DCMWC) - 1;
+		uint32_t imask = BIT(CxL1CCAP_ICMWC) - 1;
+		uint32_t waymask = (imask << 8) | dmask;
+
+		CxL1CCFG = waymask;
+		while (((CxL1CCFG >> 16) & waymask) != waymask) {
+		}
+
+		/* Prefetcher also power gates, same interface */
+		CxL1PCFG = 1;
+		while ((CxL1PCFG & 0x10000) == 0) {
+		}
+	}
+
+	/* Now set up the Xtensa CPU to enable the cache logic.  The
+	 * details of the fields are somewhat complicated, but per the
+	 * ISA ref: "Turning on caches at power-up usually consists of
+	 * writing a constant with bits[31:8] all 1’s to MEMCTL.".
+	 * Also set bit 0 to enable the LOOP extension instruction
+	 * fetch buffer.
+	 */
+#ifdef XCHAL_HAVE_ICACHE_DYN_ENABLE
+	reg = 0xffffff01;
+	__asm__ volatile("wsr %0, MEMCTL; rsync" :: "r"(reg));
+#endif
+
+	/* Likewise enable prefetching.  Sadly these values are not
+	 * architecturally defined by Xtensa (they're just documented
+	 * as priority hints), so this constant is just copied from
+	 * SOF for now.  If we care about prefetch priority tuning
+	 * we're supposed to ask Cadence I guess.
+	 */
+	reg = IS_ENABLED(CONFIG_SOC_SERIES_INTEL_CAVS_V25) ? 0x1038 : 0;
+	__asm__ volatile("wsr %0, PREFCTL; rsync" :: "r"(reg));
+
+	/* Finally we need to enable the cache in the Region
+	 * Protection Option "TLB" entries.  The hardware defaults
+	 * have this set to RW/uncached (2) everywhere.  We want
+	 * writeback caching (4) in the sixth mapping (the second of
+	 * two RAM mappings) and to mark all unused regions
+	 * inaccessible (15) for safety.  Note that there is a HAL
+	 * routine that does this (by emulating the older "cacheattr"
+	 * hardware register), but it generates significantly larger
+	 * code.
+	 */
+	if (IS_ENABLED(CONFIG_SOC_SERIES_INTEL_CAVS_V25)) {
+		/* Already set up by the ROM on older hardware. */
+		const uint8_t attribs[] = { 2, 15, 15, 15, 2, 4, 15, 15 };
+
+		for (int region = 0; region < 8; region++) {
```

Comment:
I am surprised to see a hardcoded value of 8 here. Granted, it makes complete sense--it matches the number of elements in <attribs>. I just expected to see something like ...

    for (int region = 0; region < NUM_ELEMENTS(attribs); region++) {

... however, I am unable to find such a macro (perhaps I just missed it by another name).

Anyway, this most definitely is not a show stopper; it is just an observation that jumped out at me.

Reply:
Actually this goes away in this PR.  What you're looking at here is just code motion from already-existing code, the newer API is a little cleaner and puts the region attributes into a single line of macro arguments.

But in this case the 8 is firmly architectural.  Xtensa Region Protection Option partitions memory into 8 segments based on the top three bits of the address, and that's all it does.  2^3 == 8.